### PR TITLE
[DO_NOT_MERGE][3.x]Allow substitution in parameter value field

### PR DESCRIPTION
**Pull request by mfojtik**
This allows to do following in parameters:

```json
    {
      "name": "SLAVE_SERVICE_SUFFIX",
      "description": "Random suffix of redis slave service",
      "generate": "expression",
      "from": "[a-zA-Z0-9]{3}"
    },
    {
      "name": "SLAVE_SERVICE_NAME",
      "description": "Slave Service name",
      "value": "redis-slave-${SLAVE_SERVICE_SUFFIX}"
    }
```

### CODE REVIEW

**Comment by mfojtik** [comment]
[test]

**Comment by mfojtik** [comment]
Fixes #2778

**Comment by mfojtik** [comment]
@soltysh @bparees @smarterclayton PTAL

File: pkg/template/template.go [hunk 1]
```diff
@@ -191,6 +193,29 @@ func GetParameterByName(t *api.Template, name string) *api.Parameter {
...
+		result[param.Name] = param.Value
+	}
+	return result
+}
+
+// SubstituteParameterValues substitutes all expressions in the parameters values
+func (p *Processor) SubstituteParameterValues(template *api.Template) {
+	paramMap := paramsToMap(template.Parameters)
+	for i := range template.Parameters {
+		for _, match := range parameterExp.FindAllStringSubmatch(template.Parameters[i].Value, -1) {
```
Comment (liggitt) [inline pkg/template/template.go]: need to guard against cycles

**Comment by smarterclayton** [comment]
Why is this necessary?

File: pkg/template/template.go (hunk 1 above)
Comment (liggitt) [inline pkg/template/template.go]: with good tests to make sure these don't explode:
```
FOO:"${FOO}"
```
```
FOO:"${BAR}"
BAR:"${BAZ}"
BAZ:"${BAR}"
```

**Comment by smarterclayton** [comment]
Is there another use case for this beyond James' request?

**Comment by smarterclayton** [comment]
If we do this it should be required to do this in order, and should follow the same rules as pod env replacement (you have to leave values you don't recognize).

**Comment by mfojtik** [comment]
@smarterclayton having common prefixes/suffixes in parameter values? avoid complex string values composed from multiple parameters?

**Comment by mfojtik** [comment]
@smarterclayton i'm OK with doing this post-30

**Comment by liggitt** [comment]
yeah... needs thinking through interactions with clients like the UI... what does the UI show for fields like that?

**Comment by mfojtik** [comment]
@smarterclayton AFAIK it works that way now.. if you reference parameter that is unknown it is ignored

**Comment by mfojtik** [comment]
@liggitt you show the parameter reference "foo-${bar}", is that bad?

**Comment by liggitt** [comment]
exposes it, which is ugly, and makes it vulnerable to fat-fingering by the user. also means the UI HAS to surface the names, instead of using nice display names for params if we add those

**Comment by mfojtik** [comment]
@liggitt agree. But I would still want to have this for things that requires a lot of parameters.

**Comment by smarterclayton** [comment]
Given the impact this is post 3.0.

**Comment by smarterclayton** [comment]
Let's have the discussion about this in #2778 and then once we have a decision we can potentially reopen this.

**Pull request closed without merging**